src/rust/engine/fs/src/gitignore.rs

```diff
@@ -0,0 +1,259 @@
+// Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+
+use ignore::gitignore::{Gitignore, GitignoreBuilder};
+use lazy_static::lazy_static;
+
+use crate::Stat;
+
+lazy_static! {
+  static ref EMPTY_IGNORE: Arc<GitignoreStyleExcludes> = Arc::new(GitignoreStyleExcludes {
+    patterns: vec![],
+    gitignore: Gitignore::empty(),
+  });
+}
+
+#[derive(Debug)]
+pub struct GitignoreStyleExcludes {
+  patterns: Vec<String>,
+  gitignore: Gitignore,
+}
+
+impl GitignoreStyleExcludes {
+  pub fn create(patterns: Vec<String>) -> Result<Arc<Self>, String> {
+    Self::create_with_gitignore_files(patterns, vec![])
+  }
+
+  pub fn empty() -> Arc<Self> {
+    EMPTY_IGNORE.clone()
+  }
+
+  /// Create with patterns and possibly multiple files.
+  ///
+  /// Later paths in `gitignore_paths` take precedence. `patterns` takes precedence over all
+  /// `gitignore_paths`.
+  pub fn create_with_gitignore_files(
+    patterns: Vec<String>,
+    gitignore_paths: Vec<PathBuf>,
+  ) -> Result<Arc<Self>, String> {
+    if patterns.is_empty() && gitignore_paths.is_empty() {
+      return Ok(EMPTY_IGNORE.clone());
+    }
+
+    let mut ignore_builder = GitignoreBuilder::new("");
+
+    for path in gitignore_paths {
+      if let Some(err) = ignore_builder.add(&path) {
+        return Err(format!("Error adding the path {}: {err:?}", path.display()));
+      }
+    }
+    for pattern in &patterns {
+      ignore_builder
+        .add_line(None, pattern)
+        .map_err(|e| format!("Could not parse glob exclude pattern `{pattern:?}`: {e:?}"))?;
+    }
+
+    let gitignore = ignore_builder
+      .build()
+      .map_err(|e| format!("Could not build ignore patterns: {e:?}"))?;
+
+    Ok(Arc::new(Self {
+      patterns,
+      gitignore,
+    }))
+  }
+
+  /// Return the absolute file paths to the global gitignore, `<repo>/.gitignore`, and
+  /// `<repo>/.git/info/exclude`, in that order.
+  ///
+  /// Will only add the files if they exist.
+  pub fn gitignore_file_paths(build_root: &Path) -> Vec<PathBuf> {
+    let mut result = vec![];
+
+    if let Some(global_ignore_path) =
+      ignore::gitignore::gitconfig_excludes_path().filter(|fp| fp.is_file())
```

Comment:
Does `is_file` account for symlinks?

Reply:
Yes, from the Rust docs

> Returns true if the path exists on disk and is pointing at a regular file.
>
> This function will traverse symbolic links to query information about the destination file.
>
> If you cannot access the metadata of the file, e.g. because of a permission error or broken symbolic links, this will return false.